File: Algoritmos de ordenamiento/cpp/09 Cocktail.cpp
#include <iostream>
using namespace std;

int main() {
    system("title Algoritmos de ordenamientos de datos en C++ / Cocktail Sort && cls");

    int cache = 0;

    int array[10] = {2, 3, 1, 923, -23, 9128, 0, 10, 20, 12};

    size_t size = sizeof(array) / sizeof(array[0]);

    cout << "\n== Estos son tus valores == \n\n";

    for ( int i = 0; i < size; i++ ) {
        cout << array[i] << "    ";
    }

    cout << "\n\n== Ordenando... ==\n\n";

    int start = 0;
    int end = size - 1;

    while (start < end) {
        bool swapped = false;

        for (int i = start; i < end; i++) {
            if (array[i] > array[i + 1]) {
                cache = array[i];
                array[i] = array[i + 1];
                array[i + 1] = cache;
                swapped = true;
            }
        }

        if (!swapped) {
            break;
        }

        end--;

        for (int i = end - 1; i >= start; i--) {
            if (array[i] > array[i + 1]) {
                cache = array[i];
                array[i] = array[i + 1];
                array[i + 1] = cache;
                swapped = true;
            }
        }

        start++;
        
        for ( int i = 0; i < size; i++ ) {
            cout << array[i] << "    ";
        }
        cout << endl;
    }

    cout << "\n== Estos son tus valores ordenados ==\n\n";

    for ( int i = 0; i < size; i++ ) {
        cout << array[i] << "    ";
    }

    cout << endl << endl;

    system("pause");
    system("cls");
}

File: Algoritmos de ordenamiento/py/09 Cocktail.py
def cocktail_sort(array):
    size = len(array)
    start = 0
    end = size - 1

    while start < end:
        swapped = False

        for i in range(start, end):
            if array[i] > array[i + 1]:
                array[i], array[i + 1] = array[i + 1], array[i]
                swapped = True

        if not swapped:
            break

        end -= 1

        for i in range(end - 1, start - 1, -1):
            if array[i] > array[i + 1]:
                array[i], array[i + 1] = array[i + 1], array[i]
                swapped = True

        start += 1

array = [2, 3, 1, 923, -23, 9128, 0, 10, 20, 12]

print("\n== Estos son tus valores == \n")
print(*array, sep="\t")

print("\n\n== Ordenando... ==\n")

cocktail_sort(array)

for i in range(len(array)):
    print(array[i], end="\t")
print()

print("\n== Estos son tus valores ordenados ==\n")
print(*array, sep="\t")

print("\n")

input("Presiona Enter para continuar...")
